docs/FoLiA-collect.1
.TH FoLiA-collect 1 "2021 jan 19"

.SH NAME
FoLiA-collect - collect ngram statistics of .tsv files produced by
.B FoLiA-stats

.SH SYNOPSIS
FoLiA-collect [options] DIR

.SH DESCRIPTION

.B FoLiA-collect
will process files produced by
.B FoLia-stats
found in DIR and store the result in DIR, or in
the directory provided with
.B -O

All files should match the extension:
.I *-gram.tsv
(for ngram values > 1 )
or
.I *list.tsv

.SH OPTIONS
.B --ngram
count
.RS
which ngram size to examine.
.RE

.B -O
prefix
.RS
use 'prefix' as a directory name to store the collected files in.
.RE

.B --hapax
.RS
Also include HAPAX-es. (default is to ignore them)
.RE

.B -t
number
.RS
Number of concurrent threads to be used by the program.
.RE

.B -v
.RS
be more verbose.
.RE

.B -V
or
.B
--version
.RS
Show VERSION
.RE

.B -h
or
.B --help
.RS
Show some help
.RE

.SH BUGS
possible

.SH AUTHORS

Ko van der Sloot

Martin Reynaert

e-mail: lamasoftware@science.ru.nl


NEWS
foliautils 0.16  07-01-2021
[Ko vd Sloot]
* requires libfolia 2.7 or above
* provenance data is better for a lot of modules
* added better checking on invalid NCnames in some modules.
* FoLiA-abby:
  - a lot of refactoring and additions to handle font/style information
* FoLiA-pm:
  - Notes are handled correctly now
  - fixed error in xlink attributes
* FoLiA-page:
  - more types of Page files are handled now
  - fixed annotation declarations
  - fixed offset calculation (due to change in FoLiA's opinion on those)
  - page number is added as a <br> node and in the metadata
  - added a --trusttokens option. This means that Word items in the Page file
    are added as Word's in the FoLiA, embedded in Sentences.
  - added a --norefs option to avoid adding references to the original texts
* FoLiA-correct:
  - make sure that the default is to run on 1 thread
  - added a --rebase-inputclass option

* FoLiA-alto:
  - the -t option was not always handled correctly

[Maarten van Gompel]
* FoLiA-benchmark: guard against compiler optimisation #48

foliautils 0.15  15-09-2020
[Maarten van Gompel]
* FoLiA-txt: check if a string is empty after normalisation (fix for #46)
[Ko vd Sloot]
* folia-correct: fix one-off error in hemp handling (when no hemp was found) #45
* some refactoring
  - centralized definition of XML_PARSER_OPTIONS
* bugfix in threading

foliautils 0.14  15-04-2020
[Martin Reynaert]
* updated man pages

[Ko vd Sloot]
* added man pages
* revised usage() in many modules
* the default seperator in FoLiA-stats is '_' now
* fix for: https://github.com/LanguageMachines/foliautils/issues/37
* fix for: https://github.com/LanguageMachines/foliautils/issues/41
* adapted to changes in libfolia
* many small code refactorings
* FoLiA-correct is improved a lot, allowing ngram corrections in FoLiA
* FoLiA-stats accepts a 'word_in_doc' mode now
* FoLiA-alto by default ceated <w> nodes now. use --oldstring to get <str>
* improved a lot in tests/
* many small fixes

foliautils 0.13  30-08-2019
[Ko vd Sloot]
bug fixes:
* fix for https://github.com/LanguageMachines/foliautils/issues/35
* fix for https://github.com/LanguageMachines/foliautils/issues/36

[Maarten van Gompel]
new features:
* FoLiA-wordtranslate.cxx: frog should be able to deal with
  spaces now, no need for ugly non-breaking space hack that now causes
  other problems in frog (LanguageMachines/frog#77)
* FoLiA-wordtranslate.cxx: adding ability to constrain by language


foliautils 0.12  29-05-2019
[Ko vd Sloot]
Release for FoLiA 2.0.

foliautils 0.11  15-05-2019
[Ko vd Sloot]
Stabilizing release for pre FoLiA 2.0

* Updated and added some tests
* started moving common code to a separate file and build a library
  (libfoliautils)
  - hemp detection is one of them

* FoLiA-stats:
  - added possibility to read a list of directories + file-names to process
    into seperate output directories. (could be generalized to other programs)
  - better hemp detection
* FoLiA-correct:
  - use same hempo detections as FoLiA-stats
* FoLiA-abby:
  - support more flavors
* FoLiA-clean:
  - avoid removing the last remaining tekt on <w> nodes
  - cleaning of tokenization now works


foliautils 0.10  29-11-2018
[Ko vd Sloot]
  * fixed icu:namespace issues
  * added FoLiA-abby, an ABBY to FoLiA convertor
  * src/FoLiA-abby.cxx, src/FoLiA-page.cxx, src/FoLiA-pm.cxx:
     - Allow 'none' value for --prefix
  * src/FoLiA-page.cxx, src/FoLiA-hocr.cxx: fixed Alignment info
  * src/FoLiA-correct.cxx:
     - fixed a problem with correction of the last word of a trigram.
     - fix correction of paragraphs with only deeper text
     - The --rank option accepts more flavors of files
  * src/FoLiA-stats.cxx:
     - added a --detokenize option
  * several minor fixes, refactorings etc.
  * updated tests

foliautils 0.9.2  05-06-2018
[Ko vd Sloot]
Bug fix release:
 * append small prefixes to output filenames, to ALWAYS avoid names starting with
   a numeric value.
   'FPM-' for FoLiA-pm. 'FP-' for FoLiA-page, 'FH-' for FoLiA-hocr
   Can bet set witth --prefix
 * FoLiA-stats.cxx:
   - added --collect to usage() and 'man' page
 * FoLiA-correct:
   - added --inputclass and --outputclass parameters (must be different)
   - Don't crash on empty text.

foliautils 0.9.1  17-05-2018
[Ko vd Sloot]
Bug fix release: the tests directory wasn't included in the distribution

foliautils 0.9  16-05-2018
[Ko vd Sloot]
* FoLiA-stats.cxx:
  - added a --collect option, to create files with all n-grams
    together
  - clearer message in FoLiA-stats when no results were found
  - extract text from deeper nodes, if needed
  - fixed out-of-bounds problem
  - now fails when every inputfile fails
* FoLiA-txt:
  - now fails when every inputfile fails
* avoid xml:id's starting with a number. Add "id-" in front.
* added more tests

[Maarten van Gompel]
* added codemeta.json


foliautils 0.8  19-02-2018
[Ko vd Sloot]
* added -R option to FoLiA-collect
* FoLiA-collect now can work in parallel (-t option)
* modernized configuration, whit better Max OSX support (including OpenMP)
* all modules end with an exit code now.
* added more tests to 'make check'
* added output of Type-Token Ratio's (also in degrees)
* several bugfixes.
* code cleanup and refactoring, some speedup too

foliautils 0.7  24-10-2017
[ko vd Sloot]
 * updated and expanded tests
 * fixed offset calculations in FoLiA-hocr, FoLiA-page.cxx
   and FoLiA-alto. We use unicode points now. (needed for folia v1.5 and above)
 * Changed 'modes' in FoLiA-stats, to be a bit more comprehensible
 * fixed problem with metadatatype when 'foreign-data' is present
 * enhanced FoLiA-clean. Still not done...
 * switched to dynamic OMP scheduling in most programs.
   (which process files with probably big differences in processing time)
 * small bugfixes.
 * general cleanup and refactoring

[Maarten van Gompel]
 * Added and improved FoLiA-wordtranslate.cxx

foliautils 0.6  04-04-2017
This is an intermediate release!!

src/FoLiA-collect.cxx
/*
  Copyright (c) 2014 - 2021
  CLST  - Radboud University
  ILK   - Tilburg University

  This file is part of foliautils

  foliautils is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 3 of the License, or
  (at your option) any later version.

  foliautils is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, see <http://www.gnu.org/licenses/>.

  For questions and suggestions, see:
      https://github.com/LanguageMachines/foliautils/issues
  or send mail to:
      lamasoftware (at ) science.ru.nl
*/

#include <getopt.h>
#include <cstdlib>
#include <string>
#include <stdexcept>
#include <map>
#include <vector>
#include <set>
#include <iostream>
#include <fstream>
#include <sstream>
#include "ticcutils/CommandLine.h"
#include "ticcutils/FileUtils.h"
#include "ticcutils/StringOps.h"
#include "config.h"
#ifdef HAVE_OPENMP
#include "omp.h"
#endif

using namespace	std;

bool verbose = false; // not yest used

void create_wf_list( const map<string, unsigned int>& wc,
		     const string& filename, unsigned int total ){
  ofstream os( filename );
  if ( !os ){
    cerr << "failed to create outputfile '" << filename << "'" << endl;
    exit(EXIT_FAILURE);
  }
  map<unsigned int, set<string> > wf;
  for ( const auto& cit : wc ){
    wf[cit.second].insert( cit.first );
  }
  unsigned int sum=0;
  unsigned int types=0;
  map<unsigned int, set<string> >::const_reverse_iterator wit = wf.rbegin();
  while ( wit != wf.rend() ){
    for ( const auto& st : wit->second ){
      sum += wit->first;
      os << st << "\t" << wit->first << "\t" << sum << "\t"
	 << 100 * double(sum)/total << endl;
      ++types;
    }
    ++wit;
  }
#pragma omp critical (log)
  {
    cout << "created WordFreq list '" << filename << "'" << endl;
    cout << " Stored " << sum << " tokens and " << types << " types, TTR= "
	 << (double)types/sum << endl;
  }
}

struct rec {
  unsigned int count;
  map<string,unsigned int> pc;
};

void create_lf_list( const map<string, unsigned int>& lc,
		     const string& filename, unsigned int total ){
  ofstream os( filename );
  if ( !os ){
    cerr << "failed to create outputfile '" << filename << "'" << endl;
    exit(EXIT_FAILURE);
  }
  map<unsigned int, set<string> > lf;
  for ( const auto& cit : lc ){
    lf[cit.second].insert( cit.first );
  }

  unsigned int sum=0;
  unsigned int types=0;
  map<unsigned int, set<string> >::const_reverse_iterator wit = lf.rbegin();
  while ( wit != lf.rend() ){
    for ( const auto& st : wit->second ){
      sum += wit->first;
      os << st << "\t" << wit->first << "\t" << sum << "\t"
	 << 100* double(sum)/total << endl;
      ++types;
    }
    ++wit;
  }
#pragma omp critical (log)
  {
    cout << "created LemmaFreq list '" << filename << "'" << endl;
    cout << " Stored " << sum << " tokens and " << types << " types, TTR= "
	 << (double)types/sum << endl;
  }
}

void create_lpf_list( const multimap<string, rec>& lpc,
		      const string& filename, unsigned int total ){
  ofstream os( filename );
  if ( !os ){
    cerr << "failed to create outputfile '" << filename << "'" << endl;
    exit(EXIT_FAILURE);
  }
  multimap<unsigned int, pair<string,string> > lpf;
  for ( const auto& cit : lpc ){
    for ( const auto& pit : cit.second.pc ){
      lpf.insert( make_pair( pit.second,
			     make_pair( cit.first, pit.first ) ) );
    }
  }
  unsigned int sum =0;
  multimap<unsigned int, pair<string,string> >::const_reverse_iterator wit = lpf.rbegin();
  unsigned int types = 0;
  while ( wit != lpf.rend() ){
    sum += wit->first;
    os << wit->second.first << " " << wit->second.second << "\t"
       << wit->first << "\t" << sum << "\t" << 100 * double(sum)/total << endl;
    ++types;
    ++wit;
  }
#pragma omp critical (log)
  {
    cout << "created LemmaPosFreq list '" << filename << "'" << endl;
    cout << " Stored " << sum << " tokens and " << types << " types, TTR= "
	 << (double)types/sum << endl;
  }
}

unsigned int fillWF( const string& fName,
		     map<string,unsigned int>& wf,
		     bool keepSingles ){
  unsigned int total = 0;
  ifstream is( fName );
  while ( is ){
    string line;
    getline( is, line );
    vector<string> parts = TiCC::split_at( line, "\t" );
    int num = parts.size();
    if ( num == 4 || num == 2 ){
      unsigned int cnt = TiCC::stringTo<unsigned int>( parts[1] );
      if ( keepSingles || cnt > 1 ){
	string word = parts[0];
	total += cnt;
	wf[word] += cnt;
      }
    }
  }
  return total;
}

unsigned int fillLF( const string& fName,
		     map<string,unsigned int>& lf,
		     bool keepSingles ){
  unsigned int total = 0;
  ifstream is( fName );
  while ( is ){
    string line;
    getline( is, line );
    vector<string> parts = TiCC::split_at( line, "\t" );
    int num = parts.size();
    if ( num == 4 || num == 2 ){
      unsigned int cnt = TiCC::stringTo<unsigned int>( parts[1] );
      if ( keepSingles || cnt > 1 ){
	string lemma = parts[0];
	total += cnt;
	lf[lemma] += cnt;
      }
    }
  }
  return total;
}

unsigned int fillLPF( const string& fName, unsigned int ng,
		      multimap<string, rec>& lpc,
		      bool keepSingles ){
  unsigned int total = 0;
  ifstream is( fName );
  while ( is ){
    string line;
    getline( is, line );
    vector<string> parts = TiCC::split_at( line, "\t" );
    unsigned int num = parts.size();
    if ( num == 2 || num == 4 ){
      unsigned int cnt = TiCC::stringTo<unsigned int>( parts[1] );
      if ( keepSingles || cnt > 1 ){
	vector<string> lp = TiCC::split_at( parts[0], " " );
	if ( lp.size() != 2*ng ){
	  cerr << "suprise. expected " << 2*ng << " parts, got " << lp.size()
	       << ". IN " << parts[0] << endl;
	  exit( EXIT_FAILURE);
	}
	string lemma;
	string pos;
	for ( size_t i=0; i < ng; ++i ){
	  lemma += lp[i];
	  pos += lp[ng+i];
	  if ( i != ng-1 ){
	    lemma += " ";
	    pos += " ";
	  }
	}
	multimap<string, rec >::iterator it = lpc.find(lemma);
	if ( it == lpc.end() ){
	  rec tmp;
	  tmp.count = cnt;
	  tmp.pc[pos]=cnt;
	  lpc.insert( make_pair(lemma,tmp) );
	}
	else {
	  it->second.count += cnt;
	  it->second.pc[pos] += cnt;
	}
	total += cnt;
      }
    }
  }
  return total;
}

void usage(){
  cerr << "Usage: [options] dir" << endl;
  cerr << "\t collect the ngram statistics of a directory containing" << endl;
  cerr << "\t files produced by FoLiA-stats." << endl;
  cerr << "\t--ngram\t Ngram count " << endl;
  cerr << "\t--hapax\t also include HAPAXes (default is don't) " << endl;
  cerr << "\t-O\t output directory." << endl;
  cerr << "\t-R\t recurse into the input directory." << endl;
  cerr << "\t-t <threads>\n\t--threads <threads> Number of threads to run on." << endl;
  cerr << "\t\t\t If 'threads' has the value \"max\", the number of threads is set to a" << endl;
  cerr << "\t\t\t reasonable value. (OMP_NUM_TREADS - 2)" << endl;
  cerr << "\t-h or --help\t this messages " << endl;
  cerr << "\t-V or --version\t show version " << endl;
  cerr << "\t-v\t verbosity " << endl;
}

int main( int argc, char *argv[] ){
  TiCC::CL_Options opts( "vVhO:t:R", "hapax,ngram:,help,version,threads:" );
  try {
    opts.init( argc, argv );
  }
  catch( TiCC::OptionError& e ){
    cerr << e.what() << endl;
    usage();
    exit( EXIT_FAILURE );
  }
  string nG = "1";
  int nGv = 1;
  int numThreads=1;
  string outDir;
  bool keepSingles = false;
  bool recurse = false;
  string value;
  if ( opts.extract( 'h' ) || opts.extract( "help" ) ){
    usage();
    exit(EXIT_SUCCESS);
  }
  if ( opts.extract( 'V' ) || opts.extract( "version" ) ){
    cerr << PACKAGE_STRING << endl;
    exit(EXIT_SUCCESS);
  }
  verbose = opts.extract( 'v' );
  recurse = opts.extract( 'R' );
  if ( opts.extract( "ngram", nG ) ){
    if ( !TiCC::stringTo( nG, nGv ) ){
      cerr << "unsupported value for --ngram (" << nG << ")" << endl;
      exit(EXIT_FAILURE);
    }
    if ( nGv <= 0 || nGv >= 10 ){
      cerr << "unsupported value for --ngram (" << nG << ")" << endl;
      exit(EXIT_FAILURE);
    }
  }
  opts.extract( 'O', outDir );
  keepSingles = opts.extract( "hapax" );
  if ( opts.extract( 't', value )
       || opts.extract( "threads", value ) ){
#ifdef HAVE_OPENMP
    if ( TiCC::lowercase(value) == "max" ){
      numThreads = omp_get_max_threads() - 2;
    }
    else if ( !TiCC::stringTo(value,numThreads) ) {
      cerr << "illegal value for -t (" << value << ")" << endl;
      exit( EXIT_FAILURE );
    }
#else
    cerr << "FoLiA-collect: OpenMP support is missing. -t option is not supported" << endl;
    exit(EXIT_FAILURE);
#endif
  }
  if ( !opts.empty() ){
    cerr << "unsupported options : " << opts.toString() << endl;
    usage();
    exit(EXIT_FAILURE);
  }

  vector<string> lfNames;
  vector<string> lpfNames;
  vector<string> wfNames;

  vector<string> fileNames = opts.getMassOpts();
  if ( fileNames.size() == 0 ){
    cerr << "missing a directory to process!" << endl;
    exit( EXIT_FAILURE );
  }
  else if ( fileNames.size() >1 ){
    cerr << "Only 1 directory may be specified!" << endl;
    exit( EXIT_FAILURE );
  }
  string name = fileNames[0];
  if ( !TiCC::isDir(name) ){
    cerr << "parameter '" << name << "' doesn't seem to be a directory" << endl;
    exit(EXIT_FAILURE);
  }
  else {
    string::size_type pos = name.rfind( "/" );
    if ( pos != string::npos )
      name.erase(pos);
    if ( outDir.empty() ){
      outDir = name;
    }
    else {
      TiCC::createPath(outDir);
    }
    cout << "Processing dir '" << name << "'" << endl;
    vector<string> filenames;
    if ( nGv > 1 )
      filenames = TiCC::searchFilesMatch( name, nG + "-gram.tsv", recurse );
    else
      filenames = TiCC::searchFilesMatch( name, "list.tsv", recurse );
    cout << "found " << filenames.size() << " files to process" << endl;
    for ( const auto& fullName : filenames ){
      string::size_type pos = fullName.find( ".lemmafreqlist" );
      if ( pos != string::npos ){
	lfNames.push_back( fullName );
      }
      pos = fullName.find( ".lemmaposfreqlist" );
      if ( pos != string::npos ){
	lpfNames.push_back( fullName );
      }
      pos = fullName.find( ".wordfreqlist" );
      if ( pos != string::npos ){
	wfNames.push_back( fullName );
      }
    }
  }

#ifdef HAVE_OPENMP
  omp_set_num_threads( numThreads );
  omp_set_nested( 1);
#endif

  cout << "start processing on " << numThreads << " threads" << endl;
#pragma omp parallel sections
  {

#pragma omp section
    {
      map<string,unsigned int> wf;
      unsigned int total = 0;

#pragma omp parallel for shared(wfNames)
      for ( unsigned int i=0; i < wfNames.size(); ++ i ){
	const auto& fName = wfNames[i];
#pragma omp critical (log)
	{
	  cout << "\twords \t" << fName << endl;
	}
	total += fillWF( fName, wf, keepSingles );
      }
      if ( wf.empty() ){
#pragma omp critical (log)
	{
	  cerr << "no WordFrequencies found " << endl;
	}
      }
      else {
#pragma omp critical (log)
	{
	  cout << "processed " << wfNames.size() << " wordfreq files." << endl;
	}
	string filename = outDir + ".wordfreqlist." + nG + "-gram.total.tsv";
	create_wf_list( wf, filename, total );
      }
    }
#pragma omp section
    {
      map<string,unsigned int> lf;
      unsigned int total = 0;
#pragma omp parallel for shared(lfNames)
      for ( unsigned int i=0; i < lfNames.size(); ++ i ){
	const auto& fName = lfNames[i];
#pragma omp critical (log)
	{
	  cout << "\tlemmas \t" << fName << endl;
	}
	total += fillLF( fName, lf, keepSingles );
      }
      if ( lf.empty() ){
#pragma omp critical (log)
	{
	  cerr << "no LemmaFrequencies found " << endl;
	}
      }
      else {
#pragma omp critical (log)
	{
	  cout << "processed " << lfNames.size() << " lemmafreq files." << endl;
	}
	string filename = outDir + ".lemmafreqlist." + nG + "-gram.total.tsv";
	create_lf_list( lf, filename, total );
      }
    }

#pragma omp section
    {
      multimap<string, rec> lpc;
      unsigned int total = 0;
#pragma omp parallel for shared(lpfNames)
      for ( unsigned int i=0; i < lpfNames.size(); ++ i ){
	const auto& fName = lpfNames[i];
#pragma omp critical (log)
	{
	  cout << "\tlemmapos \t" << fName << endl;
	}
	total += fillLPF( fName, nGv, lpc, keepSingles );
      }
      if ( lpc.empty() ){
#pragma omp critical (log)
	{
	  cerr << "no LemmaPosFrequencies found " << endl;
	}
      }
      else {
#pragma omp critical (log)
	{
	  cout << "processed " << lpfNames.size() << " lemmaposfreq files." << endl;
	}
	string filename = outDir + ".lemmaposfreqlist." + nG + "-gram.total.tsv";
	create_lpf_list( lpc, filename, total );
      }
    }
  }
  return EXIT_SUCCESS;
}
